Better mark_dynamic assertions

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95640
* #95639
* #95584
* __->__ #95566

This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Better mark_dynamic assertions

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/utils.py:
```diff
@@ -92,6 +92,7 @@ def profile_wrapper(*args, **kwargs):
 
 curr_frame = 0
 
+
```
I promise the linter did this lol

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
+    Given a tensor, source, and is_tensor flag, determine if a shape should be static.
+
+    Args:
+    tensor - the tensor to evaluate, parameters force a static shape.
```
You should specify if this has to be real tensor, or if fake tensor is allowed

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
+
+    Args:
+    tensor - the tensor to evaluate, parameters force a static shape.
+    source - an optional source, None forces a static shape
```
Do you remember when `wrap_to_fake_tensor_and_record` can be given a None source?

voznesenskym replied:
No, let me check

voznesenskym replied:
stacked a PR on top, I didn't see anything in tests, but there may be some part of CI that triggers it.

ezyang replied:
may i recommend good old fashioned code reading

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -4578,6 +4578,35 @@ def my_dyn_fn(x):
+            AssertionError,
+            "mark_dynamic on parameter self.param, parameters are always static today",
+        ):
+            torch._dynamo.optimize("eager")(e)(x)
```
What exactly is this test testing? Isn't this a semantics preserving refactor?

ezyang replied:
better error message

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
+    if type(tensor) is torch.nn.Parameter:
+        return (
+            True,
+            f"mark_dynamic on parameter {source.name()}, parameters are always static today.",
```
nit: consider returning a lambda instead, otherwise you always pay for string formatting even if string message is not used

ezyang replied:
Alternately, a more factored representation would be to return an enum from this function saying, essentially "why is this tensor static", and then have a separately printing function for that case. That also fixes the ugly `Tuple[bool, str]` signature which I also agree is ugly.

voznesenskym replied:
Enum sounds nice

ezyang reviewed (approved):
Please see comments

voznesenskym committed:
Update on "Better mark_dynamic assertions"


This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 


cc soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge -f "Flaky CI, failed with same error on https://github.com/pytorch/pytorch/pull/95261 this morning"
